nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,165 @@
+"""The Driver knows how to initialize the test environment
+and start the tests.
+"""
+
+from contextlib import contextmanager
+from typing import Iterator, Callable, List, Dict, Any, Type
+import atexit
+import os
+import ptpython.repl
+import tempfile
+
+from pathlib import Path
+from pprint import pprint
+
+from vlan import VLan
+from startcommand import NixStartScript
+from machine import Machine
+
+# for typing
+from logger import Logger
+
+
+class Driver:
+    """A handle to the driver that sets up the environment
+    and runs the tests"""
+    logger: Logger
+    vm_scripts: List[str]
+    tests: str
+    interactive: bool = False
+    keep_vm_state: bool = False
+    configure_python_repl: Callable = id
+    machine_class: Type[Machine] = Machine
+    vlan_class: Type[VLan] = VLan
+
+    def __init__(
+        self,
+        logger: Logger,
+        vm_scripts: List[str],
+        tests: str,
+        interactive: bool = False,
+        keep_vm_state: bool = False,
+        configure_python_repl: Callable = id,
+        machine_class: Type[Machine] = Machine,
+        vlan_class: Type[VLan] = VLan,
+    ):
+        """
+        Args:
+            - configure_python_repl: a function to configure ptpython.repl
+        """
+        self.log = logger
+        self.configure_python_repl = configure_python_repl
+        self.tests = tests
+        self.interactive = interactive
+
+        tmp_dir = Path(os.environ.get("TMPDIR", tempfile.gettempdir()))
+        tmp_dir.mkdir(mode=0o700, exist_ok=True)
+
+        self.vlans = [
+            vlan_class(int(nr), tmp_dir, logger.log_machinestate)
+            for nr in list(dict.fromkeys(os.environ.get("VLANS", "").split()))
+        ]
+
+        def cmd(scripts: List[str]) -> Iterator[NixStartScript]:
+            for s in scripts:
+                yield NixStartScript(s)
+
+        self.machines = [
+            machine_class(
+                start_command=cmd,
+                keep_vm_state=keep_vm_state,
+                name=cmd.machine_name,
```

Comment:
sorry but i still fail to see why it can't just be 

```suggestion
                name=machine_name(commandline)
```

without an additional class and that yielding function.

Reply:
I think I'd best refer you to the [start command domain docstring](https://github.com/NixOS/nixpkgs/pull/125418/files#diff-a850f8c1b6b1948f67017a2eaff91fd44de10465cc65c2c8a7795ad29a06024eR1-R3). I hope it's clear why the way it's implemented is imperative (not technically, but logically)

Reply:
ok. i don't agree at all that it should be structured like this. can we please revert to the command line assembly being a plain function and remove the `NixStartScript` class, too?

Reply:
Since I don't see a pareto-improvement here, I don't plan to change this and will resolve.

If you disagree, and want to convince me otherwise, I suggest we follow up on the matrix chat, ok?